### ISSUE TITLE
Make output machine parseable

### ISSUE THREAD

**Issue by joneshf** [stil4m/elm-analyse#102]
Thanks for this project, it's amazing!

The text output that is given is kind of okay for a human, but it's pretty hard to work with as a format. Right now some messages put the file and line range at the end, some don't have a range, some put the file at the start.

Can the error messages be formalized to be something a machine can parse? Many other linters do similar, and It'd be nice to make `elm-analyse` usable for CI tools and editors and the like. Checkstyle would be ideal for my situation. A compact format that was the same for every single message would also work really well.

I'm up for submitting a PR if you'd like it. Just let me know.

Some examples:
* csslint - https://github.com/CSSLint/csslint/wiki/Command-line-interface#--format
* eslint - http://eslint.org/docs/user-guide/command-line-interface#-f---format
* rubocop - http://rubocop.readthedocs.io/en/latest/formatters/
* shellcheck - https://github.com/koalaman/shellcheck/wiki/Integration#pick-the-output-format-that-makes-your-life-easier

**Comment by joneshf** [comment]
Did some looking around at the code. Looks like it's possible to subscribe to `messagesAsJson` and get a bunch of JSON blobs as the output. I can submit a PR to add an option `--json` or something to the CLI if that works.

Would you also accept a PR for a `--verbose` option that allows you to show all the logging when it's given and not log all the info if it's not given? That would also make using `elm-analyse` easier for CI stuff.

**Comment by joneshf** [comment]
Actually, I see that it's still a bit unnormalized. Most `data` are objects. Most also have  `type` and the payload is in a key that matches the value of the `type`, most also have `range`. But, there are exceptions.

@stil4m Any thoughts here?

**Comment by stil4m** [comment]
@joneshf Normalizing the data would be a good idea. Removing the `type`-key is definitely an improvement. I've started some [docs](https://stil4m.github.io/elm-analyse/#/messages/CoreArrayUsage) on the messages. Can you maybe provide a review on the JSON structures and put your feedback in this issue?

Related to the machine output, my suggestion would indeed be a `--json` flag that just prints one big JSON blob. I'm not sure if the `messagesAsJson` is a good port to subscribe to. This will miss a lot of valuable information that `elm-analyse` provides, such as the dependency graph, dependency info (in the future dependencies will be marked as *unused*, would be nice to have this in CI).

The reporting format itself I have to look into. I want to analyse how and why the libraries that you linked format as they do. Maybe look into some others as well.
One of the goals is to remain the type information so it can be parsed in another type-safe client (such as Elm in the elm-analyse UI, or [Type Script in VS Code](https://github.com/Krzysztof-Cieslak/vscode-elm/pull/145)).

The upcoming two weeks, unfortunately, I do not have that much time on my hands. If you create a PR I would be happy, my only concern is that I want to move to a more stable API that supports CI and editor integration :).

PS: If I do not respond, just ping me here or on Slack.

**Comment by stil4m** [comment]
Got some ideas on this: https://gist.github.com/stil4m/e418bf52ad2fe96e59fef97c80ba26b2

**Comment by joneshf** [comment]
@stil4m I took a look at the gist and I think I might have been unclear with my request.

I don't intend to suggest how you structure the output. Any structure of the data you make is fine for me. If it outputs it in a standardized format like junit or tap or checkstlye, that's even better. What I'm requesting is that all of the output be uniform.

For example:
https://github.com/stil4m/elm-analyse/blob/16101b1070352399e1086cdd6d3939fa0f4dc6a9/src/Analyser/Messages/Json.elm#L199-L200 makes a JSON blob like
```json
{
  "type": "UnreadableSourceFile",
  "UnreadableSourceFile": "some/file/path"
}
```
while https://github.com/stil4m/elm-analyse/blob/16101b1070352399e1086cdd6d3939fa0f4dc6a9/src/Analyser/Messages/Json.elm#L426-L431 makes a JSON blob like
```json
{
  "type": "SingleFieldRecord",
  "SingleFieldRecord": {
    "file": "some/file/path",
    "range": { ... }
  }
}
```
while https://github.com/stil4m/elm-analyse/blob/16101b1070352399e1086cdd6d3939fa0f4dc6a9/src/Analyser/Messages/Json.elm#L433-L438 makes a JSON blob like
```json
{
  "file": "some/file/path",
  "fieldName": "someFieldName",
  "ranges": []
}
```

As an aside, that last blob seems to break `elm-analyse`.

I could munge the data that is currently coming out myself. There's enough metadata to make it uniform for a build system to pick up. That's not my concern.

The reason I'm requesting uniform output from `elm-analyse` rather than munging the data myself is that I have a fear of upgrading versions if some uniformity isn't expected. At any version change, the upgrade path requires us to find out if the JSON output from `elm-analyse` is the same or not. Otherwise, our build process might break. That's not really a thing I want to push on the other people I work with.

**Comment by stil4m** [comment]
Ah, check. I see your points. In my work I have committed locally, I've already resolved the issue related to `DuplicateRecordFieldUpdate` not being serialized using `encodeTyped`.  As for your other point, `UnreadableSourceFile` was actually unused... (I really need to add a check to verify whether variables are used within the whole project or not). I removed this message type.

Thanks for your feedback!

As a result, the `value` for each message should now always be an object:

```json
{
    "id": 0,
    "status": "applicable",
    "files": [ ... ],
    "type": "CoreArrayUsage",
    "value": {
        ...
    }
}
```

I'll push my changes soon, and will probably make a release within a few days.
